### ISSUE TITLE
Combining Several Layers Of Keys

### ISSUE THREAD

**Issue by HxterNinja** [Kampfkarren/Roblox#68]
Recently I've been combining multiple layers of keys to the point I was getting an error. But all that change when I attacked the DataStore2 module.
For example
```lua
DataStore2.Combine("PlayerData", "PlayerAbilities")
DataStore2.Combine("PlayerAbilities", "Water")
DataStore2.Combine("Water", "Water1", "Water2")
DataStore2.Combine("Water1", "Water1Owned", "Water1Shards", "Water1Upgrades")
DataStore2.Combine("Water2", "Water2Owned", "Water2Shards", "Water2Upgrades")
```
Now when I went and set up the data for Water1Owned through Water2Upgrades, upon closing the game DataStore2 would try to save the values but hit me with an **ERROR** : "Error on BeforeSave: ServerStorage.DataStore2:583: bad argument #1 to 'pairs' (table expected, got number)".

This was due to the function on line 560 called DataStore2:__call( dataStoreName, player ). When it reached time to call dataStore:BeforeSave( function( combinedData ) ) on line 567, it would iterate through all the keys and their sub keys, but upon reaching a key that actually stored a value instead of other keys the error above would happen. 

Now if you take a look at line 574, value = combinedStore.combinedBeforeSave( clone( value ) ), what I figured out was that 
```lua
local function clone(value)
	if typeof(value) == "table" then
		return TableUtil.clone(value)
	else
		return value
	end
end
```
Would return the value stored in the value that was passed to the clone() funciton, and combinedStore.combinedBeforeSave() needed that value passed to store a table. But when you call clone( value ), and the value stored is anything but a table, it would still return the value stored and call the function with the returned value. This in turn causes the error.

On to how to fix it.

Literally all that is needed is to check whether the value being sent to clone( value ) is a table by surrounding the aforementioned line of code: value = combinedStore.combinedBeforeSave(clone(value)) on line 574 with an if statement that checks if the value is a table. Thus resulting in

```lua
if value ~= nil then
	if combinedStore.combinedBeforeSave then
		if typeof(value) == "table" then
			value = combinedStore.combinedBeforeSave(clone(value))
		end
	end
						
	combinedData[key] = value
end
```

If you are wondering what the entire function would look like:

```lua
function DataStore2:__call(dataStoreName, player)
	assert(typeof(dataStoreName) == "string" and typeof(player) == "Instance", ("DataStore2() API call expected {string dataStoreName, Instance player}, got {%s, %s}"):format(typeof(dataStoreName), typeof(player)))
	if DataStoreCache[player] and DataStoreCache[player][dataStoreName] then
		return DataStoreCache[player][dataStoreName]
	elseif combinedDataStoreInfo[dataStoreName] then
		local dataStore = DataStore2(combinedDataStoreInfo[dataStoreName], player)
		
		dataStore:BeforeSave(function(combinedData)
			
			for key in pairs(combinedData) do
				if combinedDataStoreInfo[key] then
					local combinedStore = DataStore2(key, player)
					local value = combinedStore:Get(nil, true)
					if value ~= nil then
						if combinedStore.combinedBeforeSave then
							if typeof(value) == "table" then
								value = combinedStore.combinedBeforeSave(clone(value))
							end
						end
						combinedData[key] = value
					end
				end
			end

			return combinedData
		end)

		local combinedStore = setmetatable({
			combinedName = dataStoreName,
			combinedStore = dataStore
		}, {
			__index = function(self, key)
				return CombinedDataStore[key] or dataStore[key]
			end
		})

		if not DataStoreCache[player] then
			DataStoreCache[player] = {}
		end

		DataStoreCache[player][dataStoreName] = combinedStore
		return combinedStore
	end

	local dataStore = {}

	dataStore.Name = dataStoreName
	dataStore.UserId = player.UserId

	dataStore.callbacks = {}
	dataStore.beforeInitialGet = {}
	dataStore.afterSave = {}
	dataStore.bindToClose = {}
	dataStore.savingMethod = SavingMethods.OrderedBackups.new(dataStore)

	setmetatable(dataStore, DataStoreMetatable)

	local event, fired = Instance.new("BindableEvent"), false

	game:BindToClose(function()
		if not fired then
			event.Event:wait()
		end

		local value = dataStore:Get(nil, true)

		for _, bindToClose in pairs(dataStore.bindToClose) do
			bindToClose(player, value)
		end
	end)

	local playerLeavingConnection
	playerLeavingConnection = player.AncestryChanged:Connect(function()
		if player:IsDescendantOf(game) then return end
		playerLeavingConnection:Disconnect()
		dataStore:Save()
		event:Fire()
		fired = true

		delay(40, function() --Give a long delay for people who haven't figured out the cache :^(
			DataStoreCache[player] = nil
		end)
	end)

	if not DataStoreCache[player] then
		DataStoreCache[player] = {}
	end

	DataStoreCache[player][dataStoreName] = dataStore

	return dataStore
end
```

In conclusion, this was my way of getting around the issue of setting up of multiple nested keys. I'm not entirely sure if this is the correct way to approach this issue, but it seems to work for everything that I've tested.

**Comment by HxterNinja** [comment]
If Combined data stores must have a table as a value, please reflect this in the documentation and in the error message, otherwise the fix I recommended should solve the issue.

**Comment by Kampfkarren** [comment]
> If Combined data stores must have a table as a value, please reflect this in the documentation and in the error message, otherwise the fix I recommended should solve the issue.

Not totally sure what you mean by this. Combined data stores store any form of data just fine. It's just that combining a key you also plan on using is a weird pattern I don't see a purpose in supporting.

**Comment by HxterNinja** [comment]
It's a way to be able to access data inside of a table using DataStore2("Water1Owned", plr):Get() as well as being able to access the entire table using DataStore2("Water", plr):Get(), or even get a higher level of the table. 
```lua
print(DataStore2("Water1Owned", plr):Get())

output: "true"


local water = DataStore2("Water", plr):Get()
for i,v in pairs(water) do
        print(i)
        for j,k in pairs(v) do
                print(j,k)
        end
end

output: "Water1"
"Water1Owned true"
"Water1Shards 3"
"Water1Upgrades 1"

"Water2"
"Water2Owned true"
"Water2Shards 3"
"Water2Upgrades 0"
```
But currently if you nest enough keys the error will start to occur. Thus the lines I added fixed my issue.

**Comment by Kampfkarren** [comment]
You're welcome to open up a pull request for it.

**Comment by Kampfkarren** [comment]
Closed because of the pull request you made earlier.
